### ISSUE TITLE
Client incorrectly auto-acknowledges observer notifications

### ISSUE THREAD

**Issue by weili-jiang** [obgm/libcoap#542]
I have a use case of libcoap client code establishing an observer relationship with the server.

Sometimes, when the client code cancels the observer relationship, the cancels get lost (due to network conditions). When this happens https://tools.ietf.org/html/rfc7641#section-4.5 is supposed to kick in - i.e. the server sends CON notifications and upon these timing out or getting an RST from the client, the observer relationship is removed.

What I have noticed is that in this case, libcoap always ACKs the notification, making it impossible for the server to sensibly clean-up the relationship, leading to leaks. I have tracked the issue down to `handle_response` calling `coap_send_ack` unconditionally. As observer clients are not implemented in libcoap itself (so libcoap does not know whether it should ack or not), I'm not sure how to even go about fixing this.

I have tried sending an RST in the application code but it's too late, as the server has already handled the ACK.

**Comment by weili-jiang** [comment]
I think I can work around it by disabling the auto-ack, and sending the ack manually from the application response handler. Would you accept a PR for a boolean option stored in coap_context_t that enables/disables the auto-ack behaviour?

**Comment by mrdeep1** [comment]
I agree that the current logic does not support sending a RST instead of an ACK should the application be unwilling to accept responses for the specific token.  However, I am not convinced that (albeit optionally) moving more CoAP understanding into the application layer is a good thing to do.

If the observe cancellation is getting lost (GET with observe option set to 1), then you are also likely to get the RST dropped as well, so you still have the potential for the underlying leaks issue.  If the client sees a observe response after it has done an explicit cancel, then there is no reason as to why another explicit cancel is sent.  There is however a race condition here
````
explicit cancel sent
observe received
cancel acknowledge (no observe set in response)
````
 which may trigger an unexpected second explicit cancellation.

As explicit cancels can only be done with TCP/TLS I prefer to use that rather than passive cancelling for UDP/DTLS.

**Comment by weili-jiang** [comment]
> If the observe cancellation is getting lost (GET with observe option set to 1), then you are also likely to get the RST dropped as well, so you still have the potential for the underlying leaks issue.
Actually, if the RST from the client is dropped as well, there is no problem. In this scenario, the server's CON notification retries will simply timeout. The interaction will be removed a little later than compared to an immediate RST, but it gets resolved.

> As explicit cancels can only be done with TCP/TLS I prefer to use that rather than passive cancelling for UDP/DTLS.
My use case is unfortunately UDP, where this issue presents a problem.

I agree that handling ACK and RST in the application is not good. In every other CoAP library I have used (CoAP.net, Lobaro, aiocoap, CoAPthon), observer client functionality is handled by the library. Perhaps libcoap will move in this direction too. I will continue developing my ugly workaround in the meantime.

**Comment by mrdeep1** [comment]
> the interaction will be removed a little later than compared to an immediate RST, but it gets resolved.

Good point.

I agree that things CoAP specific should be done by libcoap.   Yes, doing Observe should ideally be totally moved into libcoap.
> As explicit cancels can only be done with TCP/TLS I prefer to use that rather than passive cancelling for UDP/DTLS.

Apologies - I meant that I always use explicit cancels for both TCP and UDP so that my code does not care as to which underlying protocol is being used.

**Comment by mrdeep1** [comment]
@weili-jiang Does #552 help you at all?

**Comment by weili-jiang** [comment]
> @weili-jiang Does #552 help you at all?

Yes, that would work. I hadn't considered adding a return value because I thought it would be an ABI breakage, but doing a bit more research, it seems like returning an `int` versus `void` isn't an issue.

I'm happy to call this issue resolved with that PR.

**Comment by obgm** [comment]
Closed as resolved by merged PR #552.
